### PR TITLE
[SPARK-43133] Scala Client DataStreamWriter Foreach support

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Add the scala client side `foreach` support.

### Why are the changes needed?

Continuation of SS Connect.

### Does this PR introduce _any_ user-facing change?

Yes. Now they can use foreach in scala client exactly as how they used it in non-connect mode

### How was this patch tested?

Unit test

Manually tested it works for both `Row` and `Int`

```
import org.apache.spark.sql.{ForeachWriter, Row} 
import java.io._ 

val filePath = "/home/wei.liu/test_foreach/output-row" 

val writer = new ForeachWriter[Row] {
    var fileWriter: FileWriter = _
  
    def open(partitionId: Long, version: Long): Boolean = {
      fileWriter = new FileWriter(filePath, true)  // true to append
      true
    }
  
    def process(row: Row): Unit = {
      fileWriter.write(row.mkString(", "))  // separate fields with commas
      fileWriter.write("\n")  // newline for each row
    }
  
    def close(errorOrNull: Throwable): Unit = {
      fileWriter.close()
    }
  } 

val df = spark.readStream .format("rate") .option("rowsPerSecond", "10") .load() 
val query = df .writeStream .foreach(writer) .outputMode("update") .start() 

assert(query.exception.isEmpty)
```

```
import org.apache.spark.sql.{ForeachWriter, Row} 
import java.io._ 

val filePath = "/home/wei.liu/test_foreach/output-int" 

val writer = new ForeachWriter[Int] {
    var fileWriter: FileWriter = _
  
    def open(partitionId: Long, version: Long): Boolean = {
      fileWriter = new FileWriter(filePath, true)  // true to append
      true
    }
  
    def process(value: Int): Unit = {
      fileWriter.write(value.toString)
      fileWriter.write("\n")  // newline for each value
    }
  
    def close(errorOrNull: Throwable): Unit = {
      fileWriter.close()
    }
  } 
val df = spark.readStream .format("rate") .option("rowsPerSecond", "10") .load() 
val query = df .selectExpr("CAST(value AS INT)") .as[Int] .writeStream .foreach(writer) .outputMode("update") .start() 
```

NOTE: below __DOESN'T__ work as of now. A ticket is filed regarding this: SPARK-43796
```
import org.apache.spark.sql.ForeachWriter 
import java.io._ 

val filePath = "/home/wei.liu/test_foreach/output-custom" 

case class MyTestClass(value: Int) {
      override def toString: String = value.toString
}

val writer = new ForeachWriter[MyTestClass] {
    var fileWriter: FileWriter = _

    def open(partitionId: Long, version: Long): Boolean = {
      fileWriter = new FileWriter(filePath, true)
      true
    }

    def process(row: MyTestClass): Unit = {
      fileWriter.write(row.toString)
      fileWriter.write("\n")
    }

    def close(errorOrNull: Throwable): Unit = {
      fileWriter.close()
    }
}

val df = spark.readStream .format("rate") .option("rowsPerSecond", "10") .load()
val query = df .selectExpr("CAST(value AS INT)") .as[MyTestClass] .writeStream .foreach(writer) .outputMode("update") .start()
```

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala [hunk 1]
```diff
@@ -202,6 +206,21 @@ final class DataStreamWriter[T] private[sql] (ds: Dataset[T]) extends Logging {
...
   }
 
+  /**
+   * Sets the output of the streaming query to be processed using the provided writer object.
+   * object. See [[org.apache.spark.sql.ForeachWriter]] for more details on the lifecycle and
+   * semantics.
+   * @since 3.5.0
+   */
+  def foreach(writer: ForeachWriter[T]): DataStreamWriter[T] = {
+    val serialized = Utils.serialize(writer)
```
Comment (WweiL) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:216]: Serialize an arbitrary class instance directly?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -2293,6 +2309,19 @@ class SparkConnectPlanner(val session: SparkSession) {
       writer.queryName(writeOp.getQueryName)
     }
 
+    if (writeOp.hasForeach) {
+      if (writeOp.getForeach.hasPythonWriter) {
+        val foreach = writeOp.getForeach.getPythonWriter
+        val pythonFcn = transformPythonForeachFunction(foreach)
+        writer.foreachPython(new PythonForeachWriter(pythonFcn, dataset.schema))
+      } else {
+        val scalaFcn = Utils.deserialize[ForeachWriter[Row]]( // TODO(wei): row? String?
```
Comment (WweiL) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2318]:
```
java.lang.ClassNotFoundException: ammonite.$sess.cmd3$Helper$$anon$1
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.apache.spark.serializer.JavaDeserializationStream$$anon$1.resolveClass(JavaSerializer.scala:71)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1986)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1850)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2160)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1667)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2405)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2329)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2187)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1667)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2405)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2329)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2187)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1667)
	at java.io.ObjectInputStream.readArray(ObjectInputStream.java:2093)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1655)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2405)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2329)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2187)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1667)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2405)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2329)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2187)
```

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 3]
```diff
@@ -17,131 +17,172 @@
...
+//      // eventually(timeout(1.minute)) {
+//      // q.awaitTermination()
+//      // }
+//    }
+//  }
+
+  test("foreach") {
+    withTempPath {
+      f =>
+        val path = f.getCanonicalPath + "/output"
```
Comment (WweiL) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:153]: likely not work as foreach runs on worker

**Comment by WweiL** [comment]
This code rn is messy because it's based on unmerged PR: https://github.com/apache/spark/pull/41026

Only need to look at the part I commented

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 2 above)
Comment (WweiL) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2318]: After starting server with `./connector/connect/bin/spark-connect-shell` this is not an issue anymore.

**Comment by zhenlineo** [comment]
@WweiL I can run your test succesfully with SBT commands:
```
./build/sbt package -Phive -Pconnect
sbt "testOnly org.apache.spark.sql.streaming.StreamingQuerySuite"
```

You got the `ClassNotFound` exception probably because you were trying to run via the shell?
It would only work with Ammonite shell. If you define the user defined writer via the shell, it should also work.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * This class is shared between the client and the server to allow for serialization and
+ * deserialization of the JVM object.
+ *
+ * @param foreachWriter
+ *   The actual foreachWriter from client
+ * @param rowEncoder
+ *   An [[AgnosticEncoder]] for the input row
+ */
+@SerialVersionUID(3882541391565582579L)
+case class foreachWriterPacket(foreachWriter: AnyRef, rowEncoder: AgnosticEncoder[_])
```
Comment (rangadi) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala:36]: Why is this required? It is not not doing anything other than default serialization.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 5]
```diff
@@ -2386,10 +2393,26 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+      } else {
+        val foreachWriterPkt = unpackForeachWriter(writeOp.getForeachWriter.getScalaWriter)
+        val clientWriter = foreachWriterPkt.foreachWriter
+        if (foreachWriterPkt.rowEncoder == null) {
+          // rowEncoder is null means the client-side writer has type parameter Row,
+          // Since server-side dataset is always dataframe, here just use foreach directly.
+          writer.foreach(clientWriter.asInstanceOf[ForeachWriter[Row]])
+        } else {
+          val encoder = ExpressionEncoder(
+            foreachWriterPkt.rowEncoder.asInstanceOf[AgnosticEncoder[Any]])
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2410]: @hvanhovell is this safe? We are deserializing arbitrary Encoder class from user. Would it need user's jars? Otherwise, it could be a security issue.

Comment (rangadi) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2410]: Should we instead always convert it to row encoder based on schema?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 6]
```diff
@@ -140,10 +140,88 @@ class StreamingQuerySuite extends RemoteSparkSession with SQLHelper {
...
-      // eventually(timeout(1.minute)) {
-      // q.awaitTermination()
-      // }
+      eventually(timeout(1.minute)) {
+        q.awaitTermination()
+      }
+    }
+  }
+
+  test("foreach Row") {
```
Comment (rangadi) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:149]: Could you add one more test with custom type? Say `case class Person(name: String, id: Int)`.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 5 above)
Comment (zhenlineo) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2410]: If you could write the logic to work based on the schema of the dataset, pls do. The user's writer func would always expect a T, we need cast between row and T. It would be the best if we could not unpack the scala bytes (writer + encoder) before the execution, as the unpacking should be isolated.

However we know this issue with encoder and current impl and udfs. It is not necessary this PR's problem to address this issue.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala (hunk 4 above)
Comment (rangadi) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala:36]: Right. We can carry these two fields in the protobuf. No need to serialize them together. That way we don't need need this case class.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 5 above)
Comment (rangadi) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2410]: @zhenlineo agree. Should we try moving unpacking to execution time or doing this on server RPC thread is ok for now?

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 7]
```diff
@@ -216,6 +216,7 @@ message WriteStreamOperationStart {
 message StreamingForeachWriter {
   oneof writer {
     PythonUDF python_writer = 1;
+    ScalarScalaUDF scala_writer = 2;
```
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:219]: There was discussion about reusing `PythonUDF` or not here. We decided to reuse. Going with the same policy with Scala.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 5 above)
Comment (zhenlineo) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2410]: It is okay for now.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala (hunk 4 above)
Comment (WweiL) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala:36]: @rangadi This is basically because we also do a similar thing for UDF 
https://github.com/apache/spark/blob/master/connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/UdfPacket.scala#L39

Comment (WweiL) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala:36]: > We can carry these two fields in the protobuf
For this I think if we are reusing scalaUDF it's hard. We might could use Tuple2 here

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (WweiL) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:219]: @zhenlineo For python at least pythonVersion is needed so I think it's better to just use PythonUDF? Then I think it's probably better to use scalaUDF for same reason. 

> Would you add more fields in the UDFs or in this StreamingForeachWriter?

I don't think so. We don't seem to update foreachWriter in the future.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * This class is shared between the client and the server to allow for serialization and
+ * deserialization of the JVM object.
+ *
+ * @param foreachWriter
+ *   The actual foreachWriter from client
+ * @param rowEncoder
+ *   An [[AgnosticEncoder]] for the input row
+ */
+@SerialVersionUID(3882541391565582579L)
+case class foreachWriterPacket(foreachWriter: AnyRef, datasetEncoder: AgnosticEncoder[_])
```
Comment (rangadi) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/foreachWriterPacket.scala:33]: Fix class name, should not start with lower case.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala [hunk 9]
```diff
@@ -451,13 +451,17 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
     foreachImplementation(writer.asInstanceOf[ForeachWriter[Any]])
   }
 
-  private[sql] def foreachImplementation(writer: ForeachWriter[Any]): DataStreamWriter[T] = {
+  private[sql] def foreachImplementation(writer: ForeachWriter[Any],
+      encoder: Either[ExpressionEncoder[Any], InternalRow => Any] = null): DataStreamWriter[T] = {
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:455]: `Either` implementation is not needed. Only `Left` is used. 
Actually this change is not needed at all, I think.

**Review by rangadi — COMMENTED** [review]
Couple of comments.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala (hunk 9 above)
Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:455]: Yes, we are not supporting custom encoder yet. The unit test is disabled.

Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:455]: We can make this change when we have that working.

Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: I think we need to be cautious here. Would it be a problem if we update this in the future. And a client with old code try to connect to the new server?

Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: But we don't know if this is correct.

Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: Also I think even with only generic types the encoder is still needed. Let me test it out

Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: Confirmed, I make the encoder to be null and run the `int` code
```
Caused by: java.lang.ClassCastException: org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema cannot be cast to java.lang.Integer
```

This is because on the server, the ds is always df (Dataset[Row]), and therefore ds.expenc doesn't cast the row to Int, which the `process` method uses

**Comment by WweiL** [comment]
Getting a scala test related error, the code works fine with manual REPL
```
 - foreach Row *** FAILED *** (15 milliseconds)
[info]   java.io.NotSerializableException: org.scalatest.Engine
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
[info]   at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[info]   at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[info]   at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
[info]   at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[info]   at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[info]   at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
[info]   at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[info]   at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[info]   at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
[info]   at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
[info]   at org.apache.spark.util.Utils$.serialize(Utils.scala:126)
[info]   at org.apache.spark.sql.streaming.DataStreamWriter.foreach(DataStreamWriter.scala:226)
[info]   at org.apache.spark.sql.streaming.StreamingQuerySuite.$anonfun$new$12(StreamingQuerySuite.scala:204)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[info]   at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTest(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1564)
[info]   at org.scalatest.Suite.run(Suite.scala:1114)
[info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[info]   at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
[info]   at org.apache.spark.sql.streaming.StreamingQuerySuite.org$scalatest$BeforeAndAfterAll$$super$run(StreamingQuerySuite.scala:35)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.sql.streaming.StreamingQuerySuite.run(StreamingQuerySuite.scala:35)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:413)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:750)
```

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 10]
```diff
@@ -163,13 +165,106 @@ class StreamingQuerySuite extends RemoteSparkSession with SQLHelper {
...
-      // eventually(timeout(1.minute)) {
-      // q.awaitTermination()
-      // }
+      eventually(timeout(1.minute)) {
+        q.awaitTermination()
+      }
+    }
+  }
+
+  class TestForeachWriter[T] extends ForeachWriter[T] {
```
Comment (zhenlineo) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:174]: Can you move this class outside of the test class e.g.
```
class XXSuite {}
class TestForeachWriter[T] ... r{}
```
And see if the serdes problem is resolved.

Comment (WweiL) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: Thanks! That worked!

**Review by zhenlineo — APPROVED** [review]
The change looks good given the two limitations on the scala client:
1. RowEncoder is not sent and handled correctly.
2. No good API/example to delay the udf and encode des. 

We will let you know once the above issues have been addressed.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala [hunk 11]
```diff
@@ -202,6 +208,28 @@ final class DataStreamWriter[T] private[sql] (ds: Dataset[T]) extends Logging {
...
+  /**
+   * Sets the output of the streaming query to be processed using the provided writer object.
+   * object. See [[org.apache.spark.sql.ForeachWriter]] for more details on the lifecycle and
+   * semantics.
+   * @since 3.5.0
+   */
+  def foreach(writer: ForeachWriter[T]): DataStreamWriter[T] = {
+    // ds.encoder equal to UnboundRowEncoder means type parameter T is Row,
+    // which is not able to be serialized. Server will detect this and use default encoder.
+    val rowEncoder = if (ds.encoder != UnboundRowEncoder) {
```
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:221]: What is TODO about? There are no details. The ticket just links back to this comment. Can we have the issue described here?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 12]
```diff
@@ -2445,10 +2451,24 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+          new PythonForeachWriter(pythonFcn, dataset.schema).asInstanceOf[ForeachWriter[Any]])
+      } else {
+        val foreachWriterPkt = unpackForeachWriter(writeOp.getForeachWriter.getScalaWriter)
+        val clientWriter = foreachWriterPkt.foreachWriter
+        if (foreachWriterPkt.datasetEncoder == null) {
+          // datasetEncoder is null means the client-side writer has type parameter Row,
+          // Since server-side dataset is always dataframe, here just use foreach directly.
+          writer.foreach(clientWriter.asInstanceOf[ForeachWriter[Row]])
+        } else {
+          val encoder = ExpressionEncoder(
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2467]: Is this case possible yet un this PR? Which unit test would trigger it (may be `"forEach Int"`)?

**Review by rangadi — COMMENTED** [review]
LGTM with more description of the issue for the TODO below.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (WweiL) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:221]: The action item is item 1 in Zhen's approval note: https://github.com/apache/spark/pull/41129#pullrequestreview-1463092531

`UnboundRowEncoder` needs special care here which is not very ideal and they want to update

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 12 above)
Comment (WweiL) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2467]: Yes
